keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,890 @@
+is a common cause for user errors. Instead of having a split brain situation,
+there should be a single source of truth for the cgroup driver.
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG-Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI protocol to inform the kubelet which cgroup driver should be
+used.
+
+```diff
+ message RuntimeStatus {
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime
++    RuntimeConfiguration configuration = 2;
+ }
+ 
++message RuntimeConfiguration {
++    // Configuration information for Linux-based runtimes
++    LinuxRuntimeConfiguration linux = 1;
++}
+ 
++message LinuxRuntimeConfiguration {
++    // Cgroup driver to use
++    CgroupDriver cgroup_driver = 1;
++}
++
++enum CgroupDriver {
++    CGGROUPFS = 0;
++    SYSTEMD = 1;
++}
+```
+
+The existing RuntimeStatus message (of the existing Status API endpoint) is
+used as this is being frequently queried by the kubelet, and is a place where
+the runtime tells the Kubelet about its state. The runtime will decide which
+CgroupDriver to choose based on existing methods: its own configuration.
+
+### Kubelet
+
+Kubelet will be modified to support the new field.
+
+If available the cgroup driver information received from the container runtime
+will take precedence over cgroupDriver setting from the kubelet config (or
+`--cgroup-driver` command line flag). If the runtime does not provide
+information about the cgroup driver, then kubelet will fall back to using its
+own configuration (`cgroupDriver` from kubeletConfig or the `--cgroup-driver`
+flag). Further, the kubeletConfig field and `--cgroup-driver` flag will be
+marked as deprecated, to be dropped when the runtimes the Kubelet is supported
+to run with all support the flag.
```

Comment:
AIUI we do not publish a list of supported vs. unsupported container runtimes.

(We do tell people what we test against, but I don't think we - Kubernetes - make a public claim about compatibility with some runtimes and not others).

Reply:
True. Though in practice we only have a few implementations. We could drop the deprecation part of the proposal, as the flag is still valid and necessary with runtime versions that doesn't support this. Any thoughts @haircommander?

Reply:
We could also be explicit in what we mean by all. Members of the cri-o and containerd community have signed-off on the proposal. I am not aware of cri-dockerd's awareness. We could change the wording to
```
to be dropped when support for the field is adopted by CRI-O and containerd.
```

Reply:
tbh I think we should deprecate so users have warning that they should remove it from their scripts and the like. I don't think we need to be explicit about when we will drop it.

```
The --cgroup-driver option has been deprecated and will be dropped in a future release. Please upgrade to a CRI implementation that supports cgroup-driver detection.
```
or something

Reply:
OK, added back with the following wording:

```
Further, the kubeletConfig field and `--cgroup-driver` flag will be
marked as deprecated, to be dropped when support for the feature is adopted by
CRI-O and containerd. Usage of the deprecated setting will produce a log
message, e.g.:

"cgroupDriver option has been deprecated and will be dropped in a future release. Please upgrade to a CRI implementation that supports cgroup-driver detection."
```